src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,496 @@
+
+    @voltage_limit_high.setter
+    def voltage_limit_high(self, value):
+        self._voltage_limit_high = value
+
+    @property
+    def frequency(self):
+        return self._frequency
+
+    @frequency.setter
+    def frequency(self, value):
+        self._frequency = value
+
+    @property
+    def duty_cycle(self):
+        return self._duty_cycle
+
+    @duty_cycle.setter
+    def duty_cycle(self, value):
+        self._duty_cycle = value
+
+class Keysight33622ASimulator:
+    """
+    Simulates a Keysight 33622A waveform generator
+    """
+
+    def __init__(self, timeout=1):
+        self.timeout = timeout
+
+        self._output_buf = b""  # what the commands sends back to the "host computer"
+        self._input_buf = b""  # what we receive from the "host computer"
+
+        channel1 = OutputStates(tracking='on', voltage_low=-5.0, voltage_high=5.0, voltage_limit_low=-5.0,
+                                voltage_limit_high=5.0, frequency=40e6, duty_cycle=50)
+
+        channel2 = OutputStates(tracking='on', voltage_low=-5.0, voltage_high=5.0, voltage_limit_low = -5.0,
+                                voltage_limit_high = 5.0, frequency = 40e6, duty_cycle = 50)
+
+        self._output_states = {'channel1': channel1, 'channel2': channel2}
+
+    def sendall(self, data):
+        self._input_buf += data
+        msgs = self._input_buf.split(b"\n")
+        for m in msgs[:-1]:
+            self._parseMessage(m)  # will update _output_buf
+
+        self._input_buf = msgs[-1]
+
+    def recv(self, size=1):
+        ret = self._output_buf[:size]
+        self._output_buf = self._output_buf[len(ret):]
+
+        if len(ret) < size:
+            # simulate timeout
+            time.sleep(self.timeout)
+        return ret
+
+    def close(self):
+        # using read or write will fail after that
+        del self._output_buf
+        del self._input_buf
+
+    def _sendAnswer(self, ans):
+        self._output_buf += "%s\n" % (ans,)
+
+    def _parseMessage(self, msg):
+        """
+        msg (str): the message to parse
+        return None: self._output_buf is updated if necessary
+        """
+        logging.debug("SIM: parsing %s", to_str_escape(msg))
+        msg = msg.decode("latin1").strip()  # remove leading and trailing whitespace
+        msg = "".join(msg.split())  # remove all space characters
+
+        if msg == "*IDN?":
+            self._sendAnswer("SIMULATED Keysight Arbitrary Waveform Generator,33622A,"
+                             "FAKE_Serial_Number: AB12345678")
+        elif re.match(r"sour[1-2]:freq ", msg):
+            channel = msg[4]
+            frequency = float(msg.split()[1])
+            if frequency < 1.0E-6 or frequency > 1.0E6:
+                logging.warning("The frequency value is out of bounds: %s given", frequency)
+            else:
+                self._output_states['channel' + channel].frequency = frequency
+        elif re.match(r"sour[1-2]:trac ", msg):
```

Comment:
Please run your test cases with the simulator again. There are some (new) commands that the simulator doesn't support. You need to handle them too.